File: takengine/takengine/src/main/java/com/atakmap/opengl/GLTextureAtlas.java

package com.atakmap.opengl;

import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.SortedSet;
import java.util.TreeSet;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.RectF;
import android.opengl.GLUtils;

public class GLTextureAtlas {

    private final static Comparator<Rect> HORIZONTAL_FREE_COMPARATOR = new Comparator<Rect>() {
        @Override
        public int compare(Rect r1, Rect r2) {
            int retval = r1.area - r2.area;
            if (retval != 0)
                return retval;
            retval = r1.height - r2.height;
            if (retval != 0)
                return retval;
            return (r1.instance - r2.instance);
        }
    };

    private final static Comparator<Rect> VERTICAL_FREE_COMPARATOR = new Comparator<Rect>() {
        @Override
        public int compare(Rect r1, Rect r2) {
            int retval = r1.area - r2.area;
            if (retval != 0)
                return retval;
            retval = r1.width - r2.width;
            if (retval != 0)
                return retval;
            return (r1.instance - r2.instance);
        }
    };

    private Map<String, Long> uriToKey;
    private final int texSize;
    private int freeIndex;
    private int currentTexId;

    private Map<Long, Rect> keyToIconRect;
    private final boolean splitFreeHorizontal;
    private SortedSet<Rect> freeList;

    private final boolean fixedIconSize;
    private final int iconSize;

    private static final String TAG = "GLTextureAtlas";

    public GLTextureAtlas(int texSize) {
        this(texSize, false);
    }

    public GLTextureAtlas(int texSize, boolean splitHorizontal) {
        this(texSize, 0, splitHorizontal);
    }

    public GLTextureAtlas(int texSize, int iconSize) {
        this(texSize, iconSize, false);
    }

    private GLTextureAtlas(int texSize, int iconSize, boolean splitHorizontal) {
        this.iconSize = iconSize;
        this.texSize = 1 << (int) Math.ceil(Math.log(texSize) / Math.log(2));

        this.fixedIconSize = (iconSize > 0);

        this.uriToKey = new HashMap<String, Long>();
        this.freeIndex = 0;
        this.currentTexId = 0;

        this.splitFreeHorizontal = splitHorizontal;

        Comparator<Rect> comp;
        if (!this.fixedIconSize) {
            this.keyToIconRect = new HashMap<Long, Rect>();

            if (this.splitFreeHorizontal)
                comp = HORIZONTAL_FREE_COMPARATOR;
            else
                comp = VERTICAL_FREE_COMPARATOR;
        } else {
            this.keyToIconRect = null;
            comp = HORIZONTAL_FREE_COMPARATOR;
        }
        this.freeList = new TreeSet<Rect>(comp);
    }

    /**
     * Releases all of the textures associated with this atlas.
     */
    public void release() {
        Set<Integer> texIds = new HashSet<Integer>();
        Iterator<Long> iter = this.uriToKey.values().iterator();
        while (iter.hasNext())
            texIds.add(Integer.valueOf(this.getTexId(iter.next().longValue())));

        this.uriToKey.clear();
        if(this.keyToIconRect != null)
            this.keyToIconRect.clear();

        int[] deleteTexIds = new int[texIds.size()];
        int idx = 0;
        for (Integer texId : texIds)
            deleteTexIds[idx++] = texId.intValue();
        GLES20FixedPipeline.glDeleteTextures(idx, deleteTexIds, 0);
        
        this.freeIndex = 0;
        this.currentTexId = 0;
        this.freeList.clear();
    }

    /**
     * Releases the specified atlas texture.
     * 
     * @param textureId The texture ID
     */
    public void releaseTexture(int textureId) {
        Iterator<Long> iter = this.uriToKey.values().iterator();
        long key;
        while (iter.hasNext()) {
            key = iter.next().longValue();
            if (this.getTexId(key) == textureId) {
                iter.remove();
                this.keyToIconRect.remove(Long.valueOf(key));
            }
        }

        if (this.currentTexId == textureId) {
            this.currentTexId = 0;
            if(this.freeList != null)
                this.freeList.clear();
        }

        int[] textures = new int[] {
                textureId
        };
        GLES20FixedPipeline.glDeleteTextures(1, textures, 0);
    }

    /**
     * Returns the atlas key associated with the specified URI. If the atlas does not contain the
     * image associated with the specified URI, <code>0L</code> is returned.
     * 
     * @param uri The URI
     * @return The key associated with the specified URI, or <code>0L</code> if the atlas does not
     *         contain an entry for the URI.
     */
    public long getTextureKey(String uri) {
        final Long retval = this.uriToKey.get(uri);
        if (retval == null)
            return 0L;
        return retval.longValue();
    }

    /**
     * Returns <code>true</code> if all images in the atlas are of a fixed size, <code>false</code>
     * otherwise.
     * 
     * @return <code>true</code> if all images in the atlas are of a fixed size, <code>false</code>
     *         otherwise.
     */
    public boolean isFixedImageSize() {
        return this.fixedIconSize;
    }

    /**
     * Returns a {@link RectF} describing the bounds of the image associated with the specified key.
     * If the atlas does not contain an entry for the specified key, <code>null</code> will be
     * returned.
     * 
     * @param key The key
     * @return A {@link RectF} describing the bounds of the image associated with the specified key.
     */
    public RectF getImageRect(long key) {
        return this.getImageRect(key, false, new RectF());
    }

    /**
     * Returns a {@link RectF} describing the bounds of the image associated with the specified key.
     * If the atlas does not contain an entry for the specified key, <code>null</code> will be
     * returned.
     * 
     * @param key The key
     * @param normalized If <code>true</code>, the returned {@link RectF} will have its members
     *            normalized as texel values per the size of the texture containing the image.
     * @return A {@link RectF} describing the bounds of the image associated with the specified key.
     */
    public RectF getImageRect(long key, boolean normalized) {
        return this.getImageRect(key, normalized, new RectF());
    }

    /**
     * Returns a {@link RectF} describing the bounds of the image associated with the specified key.
     * If the atlas does not contain an entry for the specified key, <code>null</code> will be
     * returned.
     * 
     * @param key The key
     * @param normalized If <code>true</code>, the returned {@link RectF} will have its members
     *            normalized as texel values per the size of the texture containing the image.
     * @param rect An optionally pre-allocated {@link RectF} instance that will have its members set
     *            to the bounds of the image associated with the key. If <code>null</code> is
     *            specified, a new instance will be allocated and returned.
     * @return A {@link RectF} describing the bounds of the image associated with the specified key.
     */
    public RectF getImageRect(long key, boolean normalized, RectF rect) {
        if (rect == null)
            rect = new RectF();

        if (this.fixedIconSize) {
            final int index = this.getIndex(key);
            final int numIconCols = (this.texSize / this.iconSize);
            rect.top = (index / (float)numIconCols) * this.iconSize;
            rect.left = (index % numIconCols) * this.iconSize;
            rect.bottom = rect.top + this.iconSize - 1;
            rect.right = rect.left + this.iconSize - 1;
        } else {
            Rect r = this.keyToIconRect.get(Long.valueOf(key));
            if (r == null)
                return null;
            rect.set(r.x, r.y, r.x + r.width, r.y + r.height);
        }

        if (normalized) {
            rect.left /= (float) this.texSize;
            rect.top /= (float) this.texSize;
            rect.right /= (float) this.texSize;
            rect.bottom /= (float) this.texSize;
        }

        return rect;
    }

    /**
     * Returns the width of the image associated with the specified key.
     * 
     * @param key The key
     * @return The width of the image associated with the specified key or <code>0</code> if the
     *         atlas does not contain an entry for the key.
     */
    public int getImageWidth(long key) {
        if (this.fixedIconSize) {
            return this.iconSize;
        } else {
            Rect r = this.keyToIconRect.get(Long.valueOf(key));
            if (r == null)
                return 0;
            return r.width;
        }
    }

    /**
     * Returns the height of the image associated with the specified key.
     * 
     * @param key The key
     * @return The height of the image associated with the specified key or <code>0</code> if the
     *         atlas does not contain an entry for the key.
     */
    public int getImageHeight(long key) {
        if (this.fixedIconSize) {
            return this.iconSize;
        } else {
            Rect r = this.keyToIconRect.get(Long.valueOf(key));
            if (r == null)
                return 0;
            return r.height;
        }
    }

    /**
     * Returns the size of the textures used by the atlas. The value returned by this method should
     * be used to compute texel positions.
     * 
     * @return The size of the textures used by the atlas.
     */
    public int getTextureSize() {
        return this.texSize;
    }

    /**
     * Returns the texture ID containing the data for the image associated with the specified key.
     * 
     * @param key The key
     * @return The texture ID containing the data for the image associated with the specified key.
     */
    public int getTexId(long key) {
        return (int) ((key >> 32L) & 0xFFFFFFFFL);
    }

    /**
     * Returns the index for the entry containing the data for the image associated with the
     * specified key. The result of this method can be used to compute the bounds of the image for
     * fixed image size atlases.
     * 
     * @param key The key
     * @return The index for the entry containing the data for the image associated with the
     *         specified key.
     */
    public int getIndex(long key) {
        return (int) (key & 0xFFFFFFFFL);
    }

    /**
     * Returns the x-coordinate of the origin of the image in the texture.
     * 
     * @param key The key
     * @return The x-coordinate of the origin of the image in the texture or <code>0</code> if the
     *         atlas does not contain an entry for the specified key.
     */
    public int getImageTextureOffsetX(long key) {
        if (this.fixedIconSize) {
            final int index = this.getIndex(key);
            final int numIconCols = (this.texSize / this.iconSize);
            return (index % numIconCols) * this.iconSize;
        } else {
            Rect r = this.keyToIconRect.get(key);
            if (r == null)
                return 0;
            return r.x;
        }
    }

    /**
     * Returns the y-coordinate of the origin of the image in the texture.
     * 
     * @param key The key
     * @return The y-coordinate of the origin of the image in the texture or <code>0</code> if the
     *         atlas does not contain an entry for the specified key.
     */
    public int getImageTextureOffsetY(long key) {
        if (this.fixedIconSize) {
            final int index = this.getIndex(key);
            final int numIconCols = (this.texSize / this.iconSize);
            return (index / numIconCols) * this.iconSize;
        } else {
            Rect r = this.keyToIconRect.get(Long.valueOf(key));
            if (r == null)
                return 0;
            return r.y;
        }
    }

    /**
     * Adds the specified image to the atlas. The image will be resized if the atlas uses fixed size
     * images and the dimensions of the bitmap are not equal to the fixed size specified during
     * instantiation.
     * 
     * @param uri The URI associated with the image
     * @param bitmap The image data
     * @return The atlas key for the entry that was created.
     */
    public long addImage(String uri, Bitmap bitmap) {
        Bitmap icon = bitmap;
        try {
            final int iconWidth = this.fixedIconSize ? this.iconSize : bitmap.getWidth();
            final int iconHeight = this.fixedIconSize ? this.iconSize : bitmap.getHeight();
            final Bitmap.Config config = bitmap.getConfig();

            // make the icon compatible if it needs to be resized or if the
            // source config is not ARGB8888
            if(iconWidth != bitmap.getWidth() ||
               iconHeight != bitmap.getHeight() ||
               config != Bitmap.Config.ARGB_8888) {

                // XXX - don't use Bitmap.createScaledBitmap; see bug 2045
                Bitmap compatible = Bitmap.createBitmap(iconWidth, iconHeight,
                        Bitmap.Config.ARGB_8888);
                Canvas canvas = new Canvas(compatible);
                canvas.drawBitmap(icon,
                        new android.graphics.Rect(0, 0, bitmap.getWidth(), bitmap.getHeight()),
                        new android.graphics.Rect(0, 0, iconWidth, iconHeight),
                        null);
                icon = compatible;
            } else if (bitmap.getWidth() > this.texSize || bitmap.getHeight() > this.texSize) {
                throw new IllegalArgumentException("bitmap width(" + bitmap.getWidth() +") or height(" + bitmap.getHeight() + ") exeeds the texSize(" + this.texSize + ")");
            }

            // allocate a new texture if the current is filled
            if (this.fixedIconSize) {
                final int numIcons = (this.texSize / this.iconSize);
                if (this.freeIndex == (numIcons * numIcons))
                    this.currentTexId = 0;
            } else {
                SortedSet<Rect> tail = this.freeList.tailSet(new Rect(0, 0, icon.getWidth(), icon
                        .getHeight()));
                boolean haveFree = false;
                for (Rect r : tail) {
                    if (r.width >= icon.getWidth() && r.height >= icon.getHeight()) {
                        haveFree = true;
                        break;
                    }
                }
                if (!haveFree)
                    this.currentTexId = 0;
            }

            final int[] boundTexId = new int[1];
            GLES20FixedPipeline.glGetIntegerv(GLES20FixedPipeline.GL_TEXTURE_BINDING_2D, boundTexId, 0);

            if (this.currentTexId == 0) {
                int[] id = new int[1];
                GLES20FixedPipeline.glGenTextures(1, id, 0);
                if (id[0] == 0)
                    throw new RuntimeException("Failed to generate new texture id");
                this.currentTexId = id[0];

                GLES20FixedPipeline.glBindTexture(GLES20FixedPipeline.GL_TEXTURE_2D,
                        this.currentTexId);
                GLES20FixedPipeline.glTexParameterf(GLES20FixedPipeline.GL_TEXTURE_2D,
                        GLES20FixedPipeline.GL_TEXTURE_MAG_FILTER, GLES20FixedPipeline.GL_LINEAR);
                GLES20FixedPipeline.glTexParameterf(GLES20FixedPipeline.GL_TEXTURE_2D,
                        GLES20FixedPipeline.GL_TEXTURE_MIN_FILTER, GLES20FixedPipeline.GL_NEAREST);
                GLES20FixedPipeline
                        .glTexParameterf(GLES20FixedPipeline.GL_TEXTURE_2D,
                                GLES20FixedPipeline.GL_TEXTURE_WRAP_S,
                                GLES20FixedPipeline.GL_CLAMP_TO_EDGE);
                GLES20FixedPipeline
                        .glTexParameterf(GLES20FixedPipeline.GL_TEXTURE_2D,
                                GLES20FixedPipeline.GL_TEXTURE_WRAP_T,
                                GLES20FixedPipeline.GL_CLAMP_TO_EDGE);
                GLES20FixedPipeline.glTexImage2D(GLES20FixedPipeline.GL_TEXTURE_2D, 0,
                        GLES20FixedPipeline.GL_RGBA,
                        this.texSize, this.texSize, 0,
                        GLES20FixedPipeline.GL_RGBA, GLES20FixedPipeline.GL_UNSIGNED_BYTE, null);

                this.freeIndex = 0;
                if (!this.fixedIconSize) {
                    this.freeList.clear();
                    this.freeList.add(new Rect(0, 0, this.texSize, this.texSize));
                }
            } else {
                GLES20FixedPipeline.glBindTexture(GLES20FixedPipeline.GL_TEXTURE_2D,
                        this.currentTexId);
            }

            final long retval = ((long) this.currentTexId << 32L)
                    | ((long) this.freeIndex & 0xFFFFFFFFL);
            if (this.fixedIconSize) {
                final int numIconCols = (this.texSize / this.iconSize);
                final int x = (this.freeIndex % numIconCols) * this.iconSize;
                final int y = (this.freeIndex / numIconCols) * this.iconSize;

                GLUtils.texSubImage2D(GLES20FixedPipeline.GL_TEXTURE_2D, 0, x, y, icon);
            } else {
                Rect iconR = new Rect(0, 0, icon.getWidth(), icon.getHeight());
                SortedSet<Rect> tail = this.freeList.tailSet(iconR);
                Iterator<Rect> iter = tail.iterator();
                Rect free = null;
                while (iter.hasNext()) {
                    free = iter.next();
                    if (free.width >= icon.getWidth() && free.height >= icon.getHeight()) {
                        iter.remove();
                        break;
                    }
                    free = null;
                }
                // the icon is smaller than the texture and we have already made
                // sure that there is enough room to accommodate it
                if (free == null) {
                    throw new IllegalStateException("free is null");
                }
                
                // update to reflect the position of the free region
                iconR.x = free.x;
                iconR.y = free.y;
                // subdivide the free region, favoring a vertical or horizontal
                // split
                if (this.splitFreeHorizontal) {
                    if (free.width > icon.getWidth())
                        this.freeList.add(new Rect(free.x + icon.getWidth(), free.y, free.width
                                - icon.getWidth(), free.height));
                    if (free.height > icon.getHeight())
                        this.freeList.add(new Rect(free.x, free.y + icon.getHeight(), icon
                                .getWidth(), free.height - icon.getHeight()));
                } else {
                    if (free.height > icon.getHeight())
                        this.freeList.add(new Rect(free.x, free.y + icon.getHeight(), free.width,
                                free.height - icon.getHeight()));
                    if (free.width > icon.getWidth())
                        this.freeList.add(new Rect(free.x + icon.getWidth(), free.y, free.width
                                - icon.getWidth(), icon.getHeight()));
                }

                GLUtils.texSubImage2D(GLES20FixedPipeline.GL_TEXTURE_2D, 0, iconR.x, iconR.y, icon);

                this.keyToIconRect.put(Long.valueOf(retval), iconR);
            }

            this.uriToKey.put(uri, Long.valueOf(retval));

            this.freeIndex++;

            GLES20FixedPipeline.glBindTexture(GLES20FixedPipeline.GL_TEXTURE_2D, boundTexId[0]);

            return retval;
        } finally {
            if (icon != bitmap)
                icon.recycle();
        }

    }

    /**************************************************************************/

    private static class Rect {
        private static int nextInstance = 0;

        private int instance;
        public int x;
        public int y;
        public int width;
        public int height;
        public int area;

        public Rect(int x, int y, int width, int height) {
            this.x = x;
            this.y = y;
            this.width = width;
            this.height = height;
            
            this.area = this.width*this.height;

            this.instance = nextInstance++;
        }
    }
}

File: takkernel/engine/src/main/cpp/renderer/GLTextureAtlas.cpp
#include "GLTextureAtlas.h"

#include <cmath>

namespace
{

struct ScopeLocalBitmap
{
    ScopeLocalBitmap() :
        release(false)
    {
        bitmap.data = nullptr;
        bitmap.getScaledInstance = nullptr;
        bitmap.releaseData = nullptr;
    }
    ~ScopeLocalBitmap()
    {
        if(release && bitmap.data)
            bitmap.releaseData(bitmap);
    }

    atakmap::renderer::Bitmap bitmap;
    bool release;
};

}

namespace atakmap {
    namespace renderer {

        Rect::Rect(int x, int y, int width, int height, bool splitFreeHorizontal) : x(x), y(y), width(width), height(height), splitFreeHorizontal(splitFreeHorizontal)
        {
            static int nextInstance = 0;
            instance = nextInstance++;
        }


        void GLTextureAtlas::commonInit(int ts, int is, bool sh)
        {
            iconSize = is;
            texSize = 1 << (int)ceil(log(ts) / log(2));

            fixedIconSize = (iconSize > 0);

            freeIndex = 0;
            currentTexId = 0;

            splitFreeHorizontal = sh;
        }

        GLTextureAtlas::GLTextureAtlas(int texSize)
        {
            commonInit(texSize, 0, false);
        }

        GLTextureAtlas::GLTextureAtlas(int texSize, bool splitHorizontal)
        {
            commonInit(texSize, 0, splitHorizontal);
        }

        GLTextureAtlas::GLTextureAtlas(int texSize, int iconSize)
        {
            commonInit(texSize, iconSize, false);
        }

        GLTextureAtlas::~GLTextureAtlas()
        {
        }

        void GLTextureAtlas::release()
        {
            std::set<int> texIds;
            std::map<std::string, int64_t>::iterator iter;
            for (iter = uriToKey.begin(); iter != uriToKey.end(); ++iter)
                texIds.insert(getTexId(iter->second));

            uriToKey.clear();
            keyToIconRect.clear();

            std::set<int>::iterator siter;
            for (siter = texIds.begin(); siter != texIds.end(); ++siter)
                releaseTexture(*siter);
        }

        void GLTextureAtlas::releaseTexture(int textureId)
        {
            int64_t key;
            std::map<std::string, int64_t>::iterator iter;
            for (iter = uriToKey.begin(); iter != uriToKey.end(); ) {
                key = iter->second;
                if (getTexId(key) == textureId) {
                    auto tmp = iter;
                    tmp++;
                    uriToKey.erase(iter);
                    iter = tmp;
                    keyToIconRect.erase(key);
                } else {
                    iter++;
                }
            }

            if (currentTexId == textureId) {
                currentTexId = 0;
                freeList.clear();
            }

            GLuint tid = textureId;
            glDeleteTextures(1, &tid);
        }

        int64_t GLTextureAtlas::getTextureKey(std::string uri)
        {
            int64_t retval = 0L;
            std::map<std::string, int64_t>::iterator entry;
            entry = uriToKey.find(uri);
            if (entry != uriToKey.end())
                retval = entry->second;
            return retval;
        }

        bool GLTextureAtlas::isFixedImageSize()
        {
            return fixedIconSize;
        }

        math::Rectangle<float> GLTextureAtlas::getImageRect(int64_t key)
        {
            math::Rectangle<float> rect(0, 0, 0, 0);
            getImageRect(key, false, &rect);
            return rect;
        }

        math::Rectangle<float> GLTextureAtlas::getImageRect(int64_t key, bool normalized)
        {
            math::Rectangle<float> rect(0, 0, 0, 0);
            getImageRect(key, normalized, &rect);
            return rect;
        }

        math::Rectangle<float> *GLTextureAtlas::getImageRect(int64_t key, bool normalized, math::Rectangle<float> *rect)
        {
            if (rect == nullptr)
                rect = new math::Rectangle<float>(0, 0, 0, 0);

            if (fixedIconSize) {
                int index = getIndex(key);
                int numIconCols = (texSize / iconSize);
                rect->y = (float)((index / numIconCols) * iconSize);
                rect->x = (float)((index % numIconCols) * iconSize);
                rect->height = (float)iconSize;
                rect->width = (float)iconSize;
            } else {
                try {
                    Rect r = keyToIconRect.at(key);
                    rect->x = (float)r.x;
                    rect->y = (float)r.y;
                    rect->width = (float)r.width;
                    rect->height = (float)r.height;
                } catch (std::out_of_range) {
                    delete rect;
                    return nullptr;
                }
            }

            if (normalized) {
                rect->x /= (float) texSize;
                rect->y /= (float) texSize;
                rect->width /= (float) texSize;
                rect->height /= (float) texSize;
            }

            return rect;
        }


        int GLTextureAtlas::getImageWidth(int64_t key)
        {
            if (fixedIconSize) {
                return iconSize;
            } else {
                try {
                    Rect r = keyToIconRect.at(key);
                    return r.width;
                } catch (std::out_of_range) {
                    return 0;
                }
            }
        }

        int GLTextureAtlas::getImageHeight(int64_t key)
        {
            if (fixedIconSize) {
                return iconSize;
            } else {
                try {
                    Rect r = keyToIconRect.at(key);
                    return r.height;
                } catch (std::out_of_range) {
                    return 0;
                }
            }
        }

        int GLTextureAtlas::getTextureSize()
        {
            return texSize;
        }

        int GLTextureAtlas::getTexId(int64_t key)
        {
            return (int)((key >> 32L) & 0xFFFFFFFFL);
        }

        int GLTextureAtlas::getIndex(int64_t key)
        {
            return (int)(key & 0xFFFFFFFFL);
        }

        int GLTextureAtlas::getImageTextureOffsetX(int64_t key)
        {
            if (fixedIconSize) {
                int index = getIndex(key);
                int numIconCols = (texSize / iconSize);
                return (index % numIconCols) * iconSize;
            } else {
                try {
                    Rect r = keyToIconRect.at(key);
                    return r.x;
                } catch (std::out_of_range) {
                    return 0;
                }
            }
        }

        int GLTextureAtlas::getImageTextureOffsetY(int64_t key)
        {
            if (fixedIconSize) {
                int index = getIndex(key);
                int numIconCols = (texSize / iconSize);
                return (index / numIconCols) * iconSize;
            } else {
                try {
                    Rect r = keyToIconRect.at(key);
                    return r.y;
                } catch (std::out_of_range) {
                    return 0;
                }
            }
        }

        int64_t GLTextureAtlas::addImage(std::string uri, Bitmap bitmap) throw (std::invalid_argument)
        {
            ScopeLocalBitmap iconDestructor;
            Bitmap icon = bitmap;            

            if (fixedIconSize
                && (bitmap.width != iconSize || bitmap.height != iconSize)) {
                icon = bitmap.getScaledInstance(bitmap, iconSize, iconSize);

                iconDestructor.bitmap = icon;        
                iconDestructor.release = true;
            } else if (bitmap.width > texSize || bitmap.height > texSize) {
                throw std::invalid_argument("image too large compared to texture atlas");
            }

            // allocate a new texture if the current is filled
            if (fixedIconSize) {
                int numIcons = (texSize / iconSize);
                if (freeIndex == (numIcons * numIcons))
                    currentTexId = 0;
            } else {
                Rect bound(0, 0, icon.width, icon.height, splitFreeHorizontal);
                auto iter = freeList.lower_bound(bound);
                bool haveFree = false;
                while (iter != freeList.end()) {
                    Rect r = *iter;
                    if (r.width >= icon.width && r.height >= icon.height) {
                        haveFree = true;
                        break;
                    }
                    iter++;
                }
                if (!haveFree)
                    currentTexId = 0;
            }

            if (currentTexId == 0) {
                GLuint id;

                glGenTextures(1, &id);
                if (id == 0)
                    throw std::invalid_argument("Failed to generate new texture id");
                currentTexId = id;

                glBindTexture(GL_TEXTURE_2D, currentTexId);
                glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
                glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
                glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
                glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
                glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA,
                                             texSize, texSize, 0,
                                             GL_RGBA, GL_UNSIGNED_BYTE, nullptr);

                freeIndex = 0;
                if (!fixedIconSize)
                    freeList.insert(Rect(0, 0, texSize, texSize, splitFreeHorizontal));
            } else {
                glBindTexture(GL_TEXTURE_2D, currentTexId);
            }

            int64_t retval = ((int64_t) currentTexId << 32L)
                | ((int64_t) freeIndex & 0xFFFFFFFFL);
            if (fixedIconSize) {
                int numIconCols = (texSize / iconSize);
                int x = (freeIndex % numIconCols) * iconSize;
                int y = (freeIndex / numIconCols) * iconSize;

                glTexSubImage2D(GL_TEXTURE_2D, 0, x, y, icon.width, icon.height,
                                icon.format, icon.dataType, icon.data);
            } else {
                Rect iconR(0, 0, icon.width, icon.height, splitFreeHorizontal);
                auto iter = freeList.lower_bound(iconR);
                Rect free(0, 0, 0, 0, false);
                bool gotit = false;
                while (iter != freeList.end()) {
                    free = *iter;
                    auto freeIter = iter;
                    iter++;
                    if (free.width >= icon.width && free.height >= icon.height) {
                        freeList.erase(freeIter);
                        gotit = true;
                        break;
                    }
                }
                // the icon is smaller than the texture and we have already made
                // sure that there is enough room to accommodate it
                if (!gotit)
                    throw std::invalid_argument("Unexpected error in texture subdivision");
                // update to reflect the position of the free region
                iconR.x = free.x;
                iconR.y = free.y;
                // subdivide the free region, favoring a vertical or horizontal
                // split
                if (splitFreeHorizontal) {
                    if (free.width > icon.width)
                        freeList.insert(Rect(free.x + icon.width, free.y, 
                                              free.width - icon.width,
                                              free.height,
                                              splitFreeHorizontal));
                    if (free.height > icon.height)
                        freeList.insert(Rect(free.x, free.y + icon.height, 
                                               icon.width,
                                               free.height - icon.height,
                                               splitFreeHorizontal));
                } else {
                    if (free.height > icon.height)
                        freeList.insert(Rect(free.x, free.y + icon.height,
                                              free.width, 
                                              free.height - icon.height, 
                                              splitFreeHorizontal));
                    if (free.width > icon.width)
                        freeList.insert(Rect(free.x + icon.width, free.y,
                                         free.width - icon.width,
                                         icon.height, splitFreeHorizontal));
                }

                glTexSubImage2D(GL_TEXTURE_2D, 0, iconR.x, iconR.y, icon.width,
                                icon.height, icon.format, icon.dataType, icon.data);
                keyToIconRect.insert(std::pair<int64_t, Rect>(retval, iconR));
            }

            uriToKey[uri] = retval;

            freeIndex++;

            glBindTexture(GL_TEXTURE_2D, 0);

            return retval;
        }


        bool GLTextureAtlasRectComp::operator() (const Rect &x, const Rect &y) const
        {
            int retval;
            if (x.splitFreeHorizontal) {
                retval = x.height - y.height;
            } else {
                retval = x.width - y.width;
            }
            if (retval == 0)
                retval = (x.instance - y.instance);
            return retval < 0;
        }

    }
}



